src/audio/dp_queue.c

```diff
@@ -0,0 +1,328 @@
+{
+	k_spin_unlock(&dp_queue->lock, key);
+}
+
+struct sof_sink __sparse_cache *dp_queue_get_sink(struct dp_queue *dp_queue)
+{
+	return &dp_queue->sink_src_c->sink;
+}
+
+struct sof_source __sparse_cache *dp_queue_get_source(struct dp_queue *dp_queue)
+{
+	return &dp_queue->sink_src_c->source;
+}
+
+static size_t dp_queue_get_free_size(struct sof_sink __sparse_cache *sink)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	return dp_queue->free_space;
+}
+
+static int dp_queue_get_buffer(struct sof_sink __sparse_cache *sink, size_t req_size,
+			       void **data_ptr, void **buffer_start, size_t *buffer_size)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	/* dp_queue_get_free_size will return free size with adjustment for cacheline if needed */
+	if (req_size > dp_queue->ibs || req_size > dp_queue_get_free_size(sink))
+		return -ENODATA;
+
+	/* no need to lock, just reading data that may be modified by commit_buffer only */
+	*data_ptr = dp_queue->data_buffer + dp_queue->write_offset;
+	*buffer_start = dp_queue->data_buffer;
+	*buffer_size = dp_queue->data_buffer_size;
+
+	/* provided buffer is an empty space, the requestor will perform write operations only
+	 * no need to invalidate cache - will be overwritten anyway
+	 */
+	return 0;
+}
+
+static int dp_queue_commit_buffer(struct sof_sink __sparse_cache *sink, size_t commit_size)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	if (commit_size) {
+		k_spinlock_key_t key = dp_queue_lock(dp_queue);
+
+		if (dp_queue_is_shared(dp_queue)) {
+			/* a shared queue. We need to go through committed cachelines one-by-one
+			 * and if the whole cacheline is committed - writeback cache
+			 * and mark data as available for reading
+			 *
+			 * first, calculate the current and last committed cacheline
+			 * as offsets from buffer start
+			 */
+			uint32_t current_cacheline = dp_queue->write_offset /
+						     PLATFORM_DCACHE_ALIGN;
+
+			/* Last used cacheline may not be filled completely, calculate cacheline
+			 * containing 1st free byte
+			 */
+			uint32_t last_cacheline = (dp_queue->write_offset + commit_size + 1) /
+						   PLATFORM_DCACHE_ALIGN;
+			uint32_t total_num_of_cachelines = dp_queue->data_buffer_size /
+							   PLATFORM_DCACHE_ALIGN;
+			/* overlap? */
+			last_cacheline %= total_num_of_cachelines;
+
+			/* now go one by one.
+			 * if current_cacheline == last_full_cacheline - nothing to do
+			 */
+			while (current_cacheline != last_cacheline) {
+				/* writeback / invalidate */
+				uint8_t *ptr = dp_queue->data_buffer +
+						(current_cacheline * PLATFORM_DCACHE_ALIGN);
+				dcache_writeback_invalidate_region
+					((__sparse_force void __sparse_cache *)ptr,
+					PLATFORM_DCACHE_ALIGN);
+
+				/* mark data as available to read */
+				dp_queue->available_data += PLATFORM_DCACHE_ALIGN;
+				/* get next cacheline */
+				current_cacheline = (current_cacheline + 1) %
+						    total_num_of_cachelines;
+			}
```

Comment:
this whole loop should be a single call to `dcache_writeback_invalidate_region()` or 2 of them at most, if wrapping